### ISSUE TITLE
Training.{h,cpp} severe bug!

### ISSUE THREAD

**Issue by Error323** [glinscott/leela-chess#118]
I think I found a horrible bug in the [Training code](https://github.com/glinscott/leela-chess/blob/master/src/Training.cpp#L161) which duplicates many `TimeStep` points in a single chunk. That is, we keep adding `TimeStep` to the `m_data` vector and we only clear it if the nof steps from games played together equal or exceed 200 timesteps. Every `Training::dump_training` call keeps looping over previous existing `TimeStep` steps in the `m_data` vector up until we dump to gzipped chunk.

To be honest, I think the code here is very unreadable and therefore error prone. We should rewrite and unittest thoroughly, after all this is what we're training on. I'm afraid it convoluted our data..., but fortunately it seems the training algorithms are robust enough as we properly shuffle the data across many chunks.

@gcp I'm not sure this affects leelazero also, but you may want to look into it.

**Comment by jkiliani** [comment]
So this is not another bug in the V2 training code, but in the training data output from self-play games? What are the implications of this? So far, reinforcement learning worked well anyway, was this simply because improvement from a very low level was easy?

**Comment by Error323** [comment]
hmmm wait, maybe I'm wrong here... `Training::clear_training` is called in `play_one_game`.

**Comment by jkiliani** [comment]
Does the file size of the training data match what it should be from the format and the number of plies in one game?

**Comment by jkiliani** [comment]
By the way, does the client currently support saving training data for self-play games instead of deleting it after uploading to the server? This would be nice so contributors could look into these sort of things ourselves, since currently all data is stored on the server only, with no download links yet for .pgn files and training chunks.

**Comment by Error323** [comment]
Well if I understand correctly now, the current code requires all steps from 200 games in a single chunk. Let me investigate some more.

@jkiliani that's a good idea. I believe they get deleted currently.

**Comment by jkiliani** [comment]
Anyway, does this mean that the online training is currently experiencing severe problems, or how did you become aware of this in the first place?

**Comment by gcp** [comment]
This is working as intended: for transmission to the server, all data from a single game is stored in a chunk and that is uploaded. The TimeSteps are cleared between games.

At some point we checked whether the number of entries in a chunk file corresponded to the number of moves in a game.

**Comment by Error323** [comment]
Hmmmm I think @killerducky's [fix](https://github.com/glinscott/leela-chess/commit/a2c557b4214028c8d9c38606dd95c6365d10b0fa) has the sideeffect that we're storing 200 full games in a chunk instead of 200 plies. This results in chunkfiles of > 2M I don't think this was intended? Interestingly enough I don't see them on the server...

**Comment by Error323** [comment]
But indeed, my initial statement is wrong and there's no need for panic, sorry! Phew.

**Comment by Error323** [comment]
#119 deals with the 200 games being stored in a chunks.
